src/classes/DMLManager.cls

```diff
@@ -284,6 +273,44 @@ public inherited sharing class DMLManager {
 			}
 			operationRestrictedFields.put(obj.getSObjectType(),restrictedFields);
 		}
-	}	
+	}
+
+	// If the consumer desires a customized DML experience, including mock SObject unit testing:
+	// 1. Create an Apex class implementing the DMLManager.IDML interface and having the desired functionality
+	// 2. Set DMLManager.dmlOperations property to an instance of the Apex class
+	// 3. Execute the appropriate *AsUser or *AsSystem methods as usual
+	public static IDML dmlOperations = new DefaultDmlOperations();
+
+	// "Operation" suffix needed since the root DML operation name is an Apex keyword.
+	public interface IDML {
+		List<Database.DeleteResult> deleteOperation(List<SObject> recordsToDelete);
```

Comment:
So, I see now that the DefaultDmlOperations implementation differs from what we used to do (for example it uses Database.insert vs insert and returns List<Database.SaveResult> but we aren't returning the List<Database.SaveResult> from insertAsSystem. I realize that we declared the return type of insertAsSystem as type void, and you probably did this to avoid changing the signature, but it seems like we COULD change the return type from void to List<Database.SaveResult> without breaking backward compatibility, no? The DMLManager class and the methods it exposes are all public, so the signature can be changed. And since the return type was originally void, there are no integrations out there that are currently relying on the return type, so it seems like a benign change to pass back the List<Database.SaveResult> to the caller.

Reply:
@stohn777  -- per our conversation:

Let's copy the pattern in fflib_SObjectUnitOfWork which means
 * Use the exact same interface definition that we use in `fflib_SObjectUnitOfWork.IDML`
 * Implement a fully `virtual`-ized version of the class called `SimpleDML` and and have that be the default implementation
 * If we're planning to allow folks to set a different implementation in non-unit-test context, I'm less OK with having a public static property; instead, define a public static method called `setDmlImplemtation(IDML dmlImplementation)` so that we have flexibility in the future of validating the change to the implementation at runtime